### PR TITLE
Forced project folder for modules

### LINKED ISSUE #2078: Secure path generation
By default, path folder should be something like `/tmp/jhlite/xxxx`, `xxxx` can be a random uuid
So the cloud version is usable
The value `/tmp/jhlite` should be in `application.properties`

Then, we can add a "local" profile, which will disable this feature and allow the user to choose the path folder he/she wants

### PR DESCRIPTION
**Pull request by hdurix**
Fix #2078

### CODE REVIEW

File: src/main/java/tech/jhipster/lite/module/domain/ForcedProjectFolder.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  @Override
+  public boolean isValid(String folderPath) {
+    Assert.notNull("folderPath", folderPath);
+
+    return folderPath.startsWith(prefix) && !folderPath.contains("..");
+  }
+
+  @Override
+  public String generatePath() {
+    return PathUtil.appendFileSeparatorIfNeeded(prefix) + UUID.randomUUID();
```
Comment (DamnClin) [inline src/main/java/tech/jhipster/lite/module/domain/ForcedProjectFolder.java:25]: I think using Paths can be better (something like `Paths.get(prefix).resolve(UUID.randomUUID()).toString()`)

File: src/main/java/tech/jhipster/lite/module/domain/LooseProjectFolder.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+public class LooseProjectFolder implements JHipsterProjectFolderFactory {
+
+  @Override
+  public boolean isValid(String folderPath) {
+    return true;
+  }
+
+  @Override
+  public String generatePath() {
+    return PathUtil.appendFileSeparatorIfNeeded(SystemUtils.JAVA_IO_TMPDIR) + UUID.randomUUID();
```
Comment (DamnClin) [inline src/main/java/tech/jhipster/lite/module/domain/LooseProjectFolder.java:15]: Same remark, using Paths can be better

File: src/main/java/tech/jhipster/lite/module/domain/PathUtil.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package tech.jhipster.lite.module.domain;
+
+import java.io.File;
+import tech.jhipster.lite.error.domain.Assert;
+
+class PathUtil {
```
Comment (DamnClin) [inline src/main/java/tech/jhipster/lite/module/domain/PathUtil.java:6]: This should be deleted or totally reworked for something business related: building a valid path. Here it's only technical

File: src/main/java/tech/jhipster/lite/module/infrastructure/primary/InvalidProjectFolderException.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package tech.jhipster.lite.module.infrastructure.primary;
+
+import tech.jhipster.lite.error.domain.GeneratorException;
+
+public class InvalidProjectFolderException extends GeneratorException {
```
Comment (DamnClin) [inline src/main/java/tech/jhipster/lite/module/infrastructure/primary/InvalidProjectFolderException.java:5]: Can't be package private?

File: src/main/java/tech/jhipster/lite/module/infrastructure/primary/ProjectFolderConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  @Bean
+  @ConditionalOnProperty(value = "application.forced-project-folder")
+  JHipsterProjectFolderFactory forcedProjectFolder() {
+    return new ForcedProjectFolder(forcedProjectFolder);
+  }
+
+  @Bean
+  @ConditionalOnProperty(value = "application.forced-project-folder", matchIfMissing = true, havingValue = "dummy")
+  JHipsterProjectFolderFactory looseProjectFolder() {
+    return new LooseProjectFolder();
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/module/infrastructure/primary/ProjectFolderConfiguration.java:26]: I'd rename to FreeProjectFolder, Loose is a little bit strange but maybe it's just me ahaha

File: src/main/java/tech/jhipster/lite/module/infrastructure/primary/RestJHipsterModuleProperties.java [hunk 6]
```diff
@@ -17,7 +19,17 @@ class RestJHipsterModuleProperties {
...
+  public JHipsterModuleProperties toDomain(JHipsterProjectFolderFactory jHipsterProjectFolderFactory) {
+    Assert.notNull("jHipsterProjectFolderFactory", jHipsterProjectFolderFactory);
+
+    assertValidProjectFolder(jHipsterProjectFolderFactory);
+
     return new JHipsterModuleProperties(projectFolder, properties);
   }
+
+  private void assertValidProjectFolder(JHipsterProjectFolderFactory jHipsterProjectFolderFactory) {
+    if (!jHipsterProjectFolderFactory.isValid(projectFolder)) {
```
Comment (DamnClin) [inline src/main/java/tech/jhipster/lite/module/infrastructure/primary/RestJHipsterModuleProperties.java:31]: If that's your need we should update to isInvalid()

**Review by pascalgrimaud — APPROVED** [review]

File: src/main/java/tech/jhipster/lite/module/infrastructure/primary/InvalidProjectFolderException.java (hunk 4 above)
Comment (hdurix) [reply src/main/java/tech/jhipster/lite/module/infrastructure/primary/InvalidProjectFolderException.java:5]: not for now, used for modules and legacy

File: src/main/java/tech/jhipster/lite/module/infrastructure/primary/ProjectFolderConfiguration.java (hunk 5 above)
Comment (hdurix) [reply src/main/java/tech/jhipster/lite/module/infrastructure/primary/ProjectFolderConfiguration.java:26]: I'm not sure so I changed it

File: src/main/java/tech/jhipster/lite/module/domain/ForcedProjectFolder.java (hunk 1 above)
Comment (hdurix) [reply src/main/java/tech/jhipster/lite/module/domain/ForcedProjectFolder.java:25]: really nice! thanks

File: src/main/java/tech/jhipster/lite/module/domain/PathUtil.java (hunk 3 above)
Comment (hdurix) [reply src/main/java/tech/jhipster/lite/module/domain/PathUtil.java:6]: deleted

File: src/main/java/tech/jhipster/lite/module/infrastructure/primary/RestJHipsterModuleProperties.java (hunk 6 above)
Comment (hdurix) [reply src/main/java/tech/jhipster/lite/module/infrastructure/primary/RestJHipsterModuleProperties.java:31]: I wasn't sure when I coded it but yes I changed it

**Comment by pascalgrimaud** [comment]
There is 1 still one test to fix for Windows

**Comment by hdurix** [comment]
bounty claimed https://opencollective.com/generator-jhipster/expenses/89125

**Comment by pascalgrimaud** [comment]
@hdurix : approved, thanks so much for your hard work here

**Pull request merged**